### ISSUE TITLE
PG::Error: incomplete multibyte character: SELECT * FROM `sql_metrics`

### ISSUE THREAD

**Issue by yanickrochon** [jeremyevans/sequel#1957]
### Complete Description of Issue

Importing Superset SQLite database to Postgres, it fails on table "sql_metrics" with `NUMERIC(16)` data type.

### Simplest Possible Self-Contained Example Showing the Bug

I am not sure how to reproduce this. Can provide further information upon request.

### Full Backtrace of Exception (if any)
```
$ sequel -t -E -C sqlite:///home/user/.superset/superset_bak.db postgresql://user:xxxxxxx@localhost/superset
```

```
(... a few lines down...)
E, [2022-11-01T15:59:32.014577 #13264] ERROR -- : PG::Error: incomplete multibyte character: SELECT * FROM `sql_metrics`
I, [2022-11-01T15:59:32.014927 #13264]  INFO -- : (0.000214s) ROLLBACK
I, [2022-11-01T15:59:32.015090 #13264]  INFO -- : (0.000031s) ROLLBACK
Traceback (most recent call last):
	51: from /usr/local/bin/sequel:23:in `<main>'
	50: from /usr/local/bin/sequel:23:in `load'
	49: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/bin/sequel:217:in `<top (required)>'
	48: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/transactions.rb:195:in `transaction'
	47: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/connecting.rb:269:in `synchronize'
	46: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/connection_pool/threaded.rb:92:in `hold'
	45: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/transactions.rb:233:in `block in transaction'
	44: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/transactions.rb:258:in `_transaction'
	43: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/bin/sequel:218:in `block in <top (required)>'
	42: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/transactions.rb:195:in `transaction'
	41: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/connecting.rb:269:in `synchronize'
	40: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/connection_pool/threaded.rb:92:in `hold'
	39: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/transactions.rb:233:in `block in transaction'
	38: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/transactions.rb:258:in `_transaction'
	37: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/bin/sequel:221:in `block (2 levels) in <top (required)>'
	36: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/bin/sequel:221:in `each'
	35: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/bin/sequel:226:in `block (3 levels) in <top (required)>'
	34: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/actions.rb:152:in `each'
	33: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:364:in `fetch_rows'
	32: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/actions.rb:1142:in `execute'
	31: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:156:in `execute'
	30: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:209:in `_execute'
	29: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/connecting.rb:269:in `synchronize'
	28: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/connection_pool/threaded.rb:88:in `hold'
	27: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:216:in `block in _execute'
	26: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/logging.rb:43:in `log_connection_yield'
	25: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:216:in `block (2 levels) in _execute'
	24: from /var/lib/gems/2.7.0/gems/sqlite3-1.5.3-x86_64-linux/lib/sqlite3/database.rb:344:in `query'
	23: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:371:in `block in fetch_rows'
	22: from /var/lib/gems/2.7.0/gems/sqlite3-1.5.3-x86_64-linux/lib/sqlite3/resultset.rb:134:in `each'
	21: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/sqlite.rb:382:in `block (2 levels) in fetch_rows'
	20: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/actions.rb:152:in `block in each'
	19: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/bin/sequel:227:in `block (4 levels) in <top (required)>'
	18: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/shared/postgres.rb:1684:in `insert'
	17: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/shared/postgres.rb:1675:in `insert'
	16: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/actions.rb:397:in `insert'
	15: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:34:in `insert_sql'
	14: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:247:in `_insert_sql'
	13: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:1270:in `insert_values_sql'
	12: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:1280:in `_insert_values_sql'
	11: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:79:in `literal_append'
	10: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:1320:in `literal_array_append'
	 9: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:284:in `array_sql_append'
	 8: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:1140:in `expression_list_append'
	 7: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:1140:in `each'
	 6: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:1142:in `block in expression_list_append'
	 5: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/dataset/sql.rb:60:in `literal_append'
	 4: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/postgres.rb:819:in `literal_string_append'
	 3: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/database/connecting.rb:269:in `synchronize'
	 2: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/connection_pool/threaded.rb:88:in `hold'
	 1: from /var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/postgres.rb:819:in `block in literal_string_append'
/var/lib/gems/2.7.0/gems/sequel-5.61.0/lib/sequel/adapters/postgres.rb:819:in `escape_string': incomplete multibyte character (PG::Error)
```

**Comment by yanickrochon** [comment]
There is no table `sql_queries`. I have no idea why `SELECT * FROM sql_metrics` is being executed unless there is some sort of transformation.

As I wrote, this is migration is from migrating Superset.

**Comment by yanickrochon** [comment]
The table `sql_metrics` has a schema of 
```
sqlite> .schema sql_metrics
CREATE TABLE IF NOT EXISTS "sql_metrics" (
	created_on DATETIME NOT NULL, 
	changed_on DATETIME NOT NULL, 
	id INTEGER NOT NULL, 
	metric_name VARCHAR(255) NOT NULL, 
	verbose_name VARCHAR(1024), 
	metric_type VARCHAR(32), 
	table_id INTEGER, 
	expression TEXT NOT NULL, 
	description TEXT, 
	created_by_fk INTEGER, 
	changed_by_fk INTEGER, 
	d3format VARCHAR(128), 
	warning_text TEXT, 
	extra TEXT, 
	uuid NUMERIC(16), 
	PRIMARY KEY (id), 
	CONSTRAINT uq_sql_metrics_metric_name UNIQUE (metric_name, table_id), 
	CONSTRAINT uq_sql_metrics_uuid UNIQUE (uuid), 
	FOREIGN KEY(changed_by_fk) REFERENCES ab_user (id), 
	FOREIGN KEY(table_id) REFERENCES tables (id), 
	FOREIGN KEY(created_by_fk) REFERENCES ab_user (id)
);
```

And one row looks like this :

```
2020-06-23 21:38:46.238380|2020-06-23 21:38:46.256197|1|count|COUNT(*)|count|1|COUNT(*)||1|1||||��gp�I䣘�5
```

That is, the column `uuid` is a `NUMERIC(16)` type and it's value seems to be binary.

**Comment by jeremyevans** [comment]
You have non-numeric, non-UTF-8 data in a numeric column in your SQLite database.  When Sequel gets the value from SQLite, it's going to be a string that isn't UTF-8 encoded.  When Sequel tries to use that in a query to PostgreSQL, PostgreSQL raises an error.  Even if it was UTF-8 encoded, it would probably still raise an error if PostgreSQL couldn't implicitly cast it to numeric.

You should probably clean up all data in your SQLite database before trying to import it into PostgreSQL via `sequel -C`.  Or use your own custom import program.  You can copy the bin/sequel source code and modify it to handle this case.  The fact that bin/sequel does not handle this case automatically is not a bug, it is by design.

**Comment by yanickrochon** [comment]
Well, as I pointed out twice, this is not "my" database but Superset's. I did not create this, or at least I am not it's designer. And I am starting to wonder why everyone seem to point at this package as a solution to import from sqlite to postgres! Why would my Superset installation be different than someone else in that regard?

In any case, Superset has created `sql_metrics` as :

```
                                      Table "public.sql_metrics"
    Column     |            Type             | Collation | Nullable |             Default              
---------------+-----------------------------+-----------+----------+----------------------------------
 created_on    | timestamp without time zone |           | not null | 
 changed_on    | timestamp without time zone |           | not null | 
 id            | bigint                      |           | not null | generated by default as identity
 metric_name   | character varying(255)      |           | not null | 
 verbose_name  | character varying(1024)     |           |          | 
 metric_type   | character varying(32)       |           |          | 
 table_id      | bigint                      |           |          | 
 expression    | text                        |           | not null | 
 description   | text                        |           |          | 
 created_by_fk | bigint                      |           |          | 
 changed_by_fk | bigint                      |           |          | 
 d3format      | character varying(128)      |           |          | 
 warning_text  | text                        |           |          | 
 extra         | text                        |           |          | 
 uuid          | numeric(16,0)               |           |          | 
```

And running the command

```
$ sequel -t -E -C sqlite:///home/user/.superset/superset_bak.db postgresql://user:xxxxxxx@localhost/superset
```

has re-created this table in PostreSQL as

```
# \d sql_metrics
                                      Table "public.sql_metrics"
    Column     |            Type             | Collation | Nullable |             Default              
---------------+-----------------------------+-----------+----------+----------------------------------
 created_on    | timestamp without time zone |           | not null | 
 changed_on    | timestamp without time zone |           | not null | 
 id            | bigint                      |           | not null | generated by default as identity
 metric_name   | character varying(255)      |           | not null | 
 verbose_name  | character varying(1024)     |           |          | 
 metric_type   | character varying(32)       |           |          | 
 table_id      | bigint                      |           |          | 
 expression    | text                        |           | not null | 
 description   | text                        |           |          | 
 created_by_fk | bigint                      |           |          | 
 changed_by_fk | bigint                      |           |          | 
 d3format      | character varying(128)      |           |          | 
 warning_text  | text                        |           |          | 
 extra         | text                        |           |          | 
 uuid          | numeric(16,0)               |           |          | 
Indexes:
    "sql_metrics_pkey" PRIMARY KEY, btree (id)
```

Pretty similar! So I do not know why the data in the `uuid` column is represented as binary data, and I do not know how other people's data is represented in that column. All I know is that this module, after creating the table, fails at this point.

**Comment by yanickrochon** [comment]
From [this post](https://neosmart.net/blog/2018/converting-a-binary-blob-guid-to-text-in-sql/), the solution is this

```
SELECT id, substr(hex(uuid), 1, 8)
|| '-' || substr(hex(uuid), 9, 4)
|| '-' || substr(hex(uuid), 13, 4)
|| '-' || substr(hex(uuid), 17, 4)
|| '-' || substr(hex(uuid), 21, 12)
FROM sql_metrics;
```

Which correctly convert the column as 

```
1|DE13F867-70EC-49E4-A398-041B83F1A535
2|2214F0E0-8900-4E6D-98E2-692E170E75DB
3|37F1CB7A-1D66-43A4-AA37-F7D8C0135FB3
4|B941F58D-4BA0-4698-B774-CD1F474E65D2
5|A0FACE77-177E-4196-80C3-16DF0AE195D1
...
```

In other words, the data is not corrupted. I will continue to post here, even if this issue was immediately closed, just in case some other poor soul encounter the same problem.

**Comment by jeremyevans** [comment]
> And I am starting to wonder why everyone seem to point at this package as a solution to import from sqlite to postgres! 

Not sure.  Sequel's documentation is clear on bin/sequel's support for copying databases: http://sequel.jeremyevans.net/rdoc/files/doc/bin_sequel_rdoc.html#label-Copy+Databases

In any case, your issue is that SQLite doesn't enforce types, unless you are using strict tables, and even then you are limited to SQLite's very limited type support (5 types).  You are having binary data in a numeric column, which PostgreSQL doesn't support.  So it's completely expected in this case that using bin/sequel to copy the database from SQLite to PostgreSQL results in an error.

**Comment by yanickrochon** [comment]
For what it is worth, Superset creates Postgres `sql_metrics` table as such :

```
# \d sql_metrics
                                          Table "public.sql_metrics"
    Column     |            Type             | Collation | Nullable |                 Default                 
---------------+-----------------------------+-----------+----------+-----------------------------------------
 created_on    | timestamp without time zone |           |          | 
 changed_on    | timestamp without time zone |           |          | 
 id            | integer                     |           | not null | nextval('sql_metrics_id_seq'::regclass)
 metric_name   | character varying(255)      |           | not null | 
 verbose_name  | character varying(1024)     |           |          | 
 metric_type   | character varying(32)       |           |          | 
 table_id      | integer                     |           |          | 
 expression    | text                        |           | not null | 
 description   | text                        |           |          | 
 created_by_fk | integer                     |           |          | 
 changed_by_fk | integer                     |           |          | 
 d3format      | character varying(128)      |           |          | 
 warning_text  | text                        |           |          | 
 extra         | text                        |           |          | 
 uuid          | uuid                        |           |          | 
Indexes:
    "sql_metrics_pkey" PRIMARY KEY, btree (id)
    "uq_sql_metrics_metric_name" UNIQUE CONSTRAINT, btree (metric_name, table_id)
    "uq_sql_metrics_uuid" UNIQUE CONSTRAINT, btree (uuid)
Foreign-key constraints:
    "sql_metrics_changed_by_fk_fkey" FOREIGN KEY (changed_by_fk) REFERENCES ab_user(id)
    "sql_metrics_created_by_fk_fkey" FOREIGN KEY (created_by_fk) REFERENCES ab_user(id)
    "sql_metrics_table_id_fkey" FOREIGN KEY (table_id) REFERENCES tables(id)
```

So, I would conclude that the best course of action would be to simply move data without trying to re-create the schema. If your package could do that, I could fix it to make it work.

**Comment by jeremyevans** [comment]
You cannot "move data without trying to re-create the schema".  You have to create a table in PostgreSQL to import the data into.  Superset's use of `numeric(16)` as the SQLite type is the underlying problem.  If they used `blob` as the type (more appropriate as it stores binary data, not a numeric), I don't think you would have this issue, since Sequel would recognize this as a blob type, would use a PostgreSQL `bytea` column for the uuid, would retrieve the values and convert them to blobs, and would insert them into PostgreSQL as blobs.

bin/sequel would never automatically convert a `numeric(16)` column type in SQLite to `uuid` column type in PostgreSQL, because doing so makes no sense in the general case.

**Comment by yanickrochon** [comment]
@jeremyevans I came here because I found [this](https://subscription.packtpub.com/book/business-intelligence/9781788992244/2/ch02lvl1sec17/migrating-data-from-sqlite-to-postgresql) post, and I had also found the same answer from a solution on StackOverflow. 

All and all, this is my last comment. I hope anyone with the exact same problem, stumbling on this issue after a Google Search, will also find it useful; the `sequel` solution does not work for Superset migration.
